src/main/java/spoon/support/util/ModelList.java

```diff
@@ -0,0 +1,238 @@
+		list.set(index, element);
+		updateModCount();
+		return oldElement;
+	}
+
+	static void linkToParent(CtElement owner, CtElement element) {
+		if (owner.getFactory().getEnvironment().checksAreSkipped() == false && element.isParentInitialized() && element.getParent() != owner) {
+			//the `e` already has an different parent. Check if it is still linked to that parent
+			if (element.getRoleInParent() != null) {
+				throw new SpoonException("The element " + element + " is already used by another part of SpoonModel. Remove it from previous model or clone it before.");
+			}
+		}
+		element.setParent(owner);
+	}
+
+	@Override
+	public boolean contains(Object o) {
+		return list.contains(o);
+	}
+	@Override
+	public boolean isEmpty() {
+		return list.isEmpty();
+	}
+
+	@Override
+	public Object[] toArray() {
+		return list.toArray();
+	}
+
+	@Override
+	public <T> T[] toArray(T[] a) {
+		return list.toArray(a);
+	}
+
+	@Override
+	public boolean add(T e) {
+		if (e == null) {
+			return false;
+		}
+		CtElement owner = getOwner();
+		ensureModifiableList();
+		linkToParent(owner, e);
+		getModelChangeListener().onListAdd(owner, getRole(), list, e);
+		boolean result = list.add(e);
+		updateModCount();
+		onSizeChanged(list.size());
+		return result;
+	}
+
+	@Override
+	public boolean remove(Object o) {
+		if (list.isEmpty()) {
+			return false;
+		}
+		int size = list.size();
+		for (int i = 0; i < size; i++) {
+			//first do not use equals, but same
+			if (list.get(i) == o) {
+				remove(i);
+				return true;
+			}
+		}
+		int idx = list.indexOf(o);
+		if (idx >= 0) {
+			remove(idx);
+			return true;
+		}
+		return false;
+	}
+
+	@Override
+	public boolean containsAll(Collection<?> c) {
+		return list.containsAll(c);
+	}
+
+	@Override
+	public void clear() {
+		getModelChangeListener().onListDeleteAll(getOwner(), getRole(), list, new ArrayList<>(list));
+		list = CtElementImpl.emptyList();
+		modCount++;
+		onSizeChanged(list.size());
+	}
+
+	@Override
+	public boolean equals(Object o) {
+		return list.equals(o);
+	}
+
+	@Override
+	public int hashCode() {
+		return list.hashCode();
+	}
+
+	@Override
+	public void add(int index, T element) {
+		if (element == null) {
+			return;
+		}
+		CtElement owner = getOwner();
+		ensureModifiableList();
+		linkToParent(owner, element);
+		getModelChangeListener().onListAdd(owner, getRole(), list, index, element);
+		list.add(index, element);
+		updateModCount();
+		onSizeChanged(list.size());
+	}
+
+	@Override
+	public T remove(int index) {
+		T oldElement = list.get(index);
+		getModelChangeListener().onListDelete(getOwner(), getRole(), list, index, oldElement);
+		list.remove(index);
+		updateModCount();
+		onSizeChanged(list.size());
+		return oldElement;
+	}
+
+	@Override
+	public int indexOf(Object o) {
+		return list.indexOf(o);
+	}
+
+	@Override
+	public int lastIndexOf(Object o) {
+		return list.lastIndexOf(o);
+	}
+
+	private static class InternalList<T> extends ArrayList<T> {
```

Comment:
Could you document that this internalList purpose is to manage the ConcurrentModificationException through modCount, maybe linking the URL: https://docs.oracle.com/javase/7/docs/api/java/util/AbstractList.html#modCount

It will ease the future maintaining ;) 

And more generally I'm not sure why you update this value only when calling `clear` method: shouldn't you manage also the add/remove cases?

Reply:
> Could you document

ok, I will do so.

> And more generally I'm not sure why you update modCount only when calling `clear` method

I am directly changing it in `clear` method because this one changes content of internal list, but doesn't calls any method of that internal `list`. In all other cases the internal method of `list` is called so the modCount of that internal list is managed automatically. Then I just call `updateModCount` to read list.modCount and write it into this.modCount.

Reply:
Thanks for the explanation ;)